### PR TITLE
Increase nulling granularity to preserve valid data

### PR DESCRIPTION
**Pull request by Woffendm**
Encountering field parsing errors with a user-supplied schema no longer nullifies all columns in a row; instead, it only nulls out the field that failed to parse (Addresses issue #235 open since Feb 2017)

### CODE REVIEW

File: src/main/scala/com/databricks/spark/xml/util/TypeCast.scala [hunk 1]
```diff
@@ -80,22 +80,33 @@ object TypeCast {
...
+        case ByteType => castTo(value, ByteType, options)
+        case ShortType => castTo(value, ShortType, options)
+        case IntegerType => signSafeToInt(value, options)
+        case dt: DecimalType => castTo(value, dt, options)
+        case _ =>
+          sys.error(s"Failed to parse a value for data type $dataType.")
+      }
+    }
+    catch {
+      case e: Exception => {
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/util/TypeCast.scala]: We could do this:

```
case NonFatal(_) if options.permissive => null
```

Comment (Woffendm) [reply src/main/scala/com/databricks/spark/xml/util/TypeCast.scala]: Done.

**Comment by Woffendm** [comment]
I implemented your suggestion.

**Comment by Woffendm** [comment]
Any updates?

**Comment by gatorsmile** [comment]
cc @HyukjinKwon

File: src/main/scala/com/databricks/spark/xml/util/TypeCast.scala [hunk 2]
```diff
@@ -22,10 +22,11 @@ import java.util.Locale
 
 import scala.util.Try
 import scala.util.control.Exception._
-
 import org.apache.spark.sql.types._
 import com.databricks.spark.xml.XmlOptions
 
+import scala.util.control.NonFatal
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/util/TypeCast.scala]: ditto for import style - https://github.com/databricks/scala-style-guide#imports

File: src/main/scala/com/databricks/spark/xml/util/TypeCast.scala [hunk 3]
```diff
@@ -80,22 +81,26 @@ object TypeCast {
...
+        case ByteType => castTo(value, ByteType, options)
+        case ShortType => castTo(value, ShortType, options)
+        case IntegerType => signSafeToInt(value, options)
+        case dt: DecimalType => castTo(value, dt, options)
+        case _ =>
+          sys.error(s"Failed to parse a value for data type $dataType.")
+      }
+    }
+    catch {
+      case NonFatal(_) if options.permissive => null
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/xml/util/TypeCast.scala]: Hmmmmm.. IIRC Spark's JSON doesn't also allow the partial results. Also, wouldn't this fail to fill up the corrupt column?

**Comment by Woffendm** [comment]
My team is no longer using the SparkXML library for our ETL process; however, I'll continue to work on these pull requests as time allows. It may take a while though since we have upcoming deadlines for key deliverables.

**Comment by HyukjinKwon** [comment]
Fixed in #358

**Pull request closed without merging**